File: services/tools/project.py
from __future__ import annotations

import json
import subprocess
from pathlib import Path

from services.tools.registry import register_tool
from services.tools.workspace import tool_error, tool_ok, workspace_root

BUILD_TIMEOUT_SEC = 600
BUILD_OUTPUT_TAIL = 4000

TARGET_SCRIPTS: dict[str, list[str]] = {
    "weixin": ["build:weapp", "build:mp-weixin", "build:weixin", "dev:mp-weixin"],
    "alipay": ["build:alipay", "build:mp-alipay", "dev:mp-alipay"],
    "h5": ["build:h5", "build:web", "build"],
    "tt": ["build:tt", "build:mp-toutiao"],
    "baidu": ["build:swan", "build:mp-baidu"],
}


def _read_package_json(root: Path) -> dict | None:
    pkg = root / "package.json"
    if not pkg.is_file():
        return None
    try:
        data = json.loads(pkg.read_text(encoding="utf-8"))
    except json.JSONDecodeError:
        return None
    return data if isinstance(data, dict) else None


def _collect_deps(pkg: dict) -> dict[str, str]:
    deps: dict[str, str] = {}
    for key in ("dependencies", "devDependencies", "peerDependencies"):
        raw = pkg.get(key) or {}
        if isinstance(raw, dict):
            deps.update({str(k): str(v) for k, v in raw.items()})
    return deps


def _detect_signals(root: Path, deps: dict[str, str]) -> list[str]:
    signals: list[str] = []
    if any(k.startswith("@tarojs/") for k in deps):
        signals.append("taro")
    if (root / "config" / "index.ts").is_file() or (root / "config" / "index.js").is_file():
        if "taro" not in signals and any("taro" in k.lower() for k in deps):
            signals.append("taro")
    if any(k.startswith("@dcloudio/") or "uni-app" in k for k in deps):
        signals.append("uni-app")
    if (root / "pages.json").is_file() or (root / "manifest.json").is_file():
        if "uni-app" not in signals:
            signals.append("uni-app-like")
    if any("morjs" in k.lower() for k in deps) or (root / "mor.config.js").is_file():
        signals.append("morjs")
    if (root / "project.config.json").is_file() or (root / "project.private.config.json").is_file():
        signals.append("wechat-miniprogram-config")
    return signals


def _package_manager(root: Path) -> str:
    if (root / "pnpm-lock.yaml").is_file():
        return "pnpm"
    if (root / "yarn.lock").is_file():
        return "yarn"
    return "npm"


def _run_script(root: Path, script_name: str) -> subprocess.CompletedProcess[str]:
    pm = _package_manager(root)
    if pm == "pnpm":
        cmd = ["pnpm", "run", script_name]
    elif pm == "yarn":
        cmd = ["yarn", script_name]
    else:
        cmd = ["npm", "run", script_name]
    return subprocess.run(
        cmd,
        cwd=root,
        capture_output=True,
        text=True,
        timeout=BUILD_TIMEOUT_SEC,
        check=False,
    )


def _pick_build_script(scripts: dict[str, str], target: str) -> str | None:
    for name in TARGET_SCRIPTS.get(target, []) + ["build"]:
        if name in scripts:
            return name
    return None


@register_tool()
def detect_framework() -> str:
    """检测工作区项目使用的多端框架（uni-app/Taro/Morjs 等）及可用构建脚本。"""
    root = workspace_root()
    pkg = _read_package_json(root)
    if pkg is None:
        return tool_error("未找到可解析的 package.json")

    deps = _collect_deps(pkg)
    signals = _detect_signals(root, deps)
    scripts = pkg.get("scripts") or {}
    script_names = list(scripts.keys()) if isinstance(scripts, dict) else []

    framework = "unknown"
    if "taro" in signals:
        framework = "taro"
    elif "uni-app" in signals or "uni-app-like" in signals:
        framework = "uni-app"
    elif "morjs" in signals:
        framework = "morjs"
    elif signals:
        framework = signals[0]

    return tool_ok(
        framework=framework,
        signals=signals,
        name=pkg.get("name"),
        package_manager=_package_manager(root),
        scripts=script_names,
        suggested_build={
            target: _pick_build_script(scripts, target) if isinstance(scripts, dict) else None
            for target in TARGET_SCRIPTS
        },
    )


@register_tool()
def run_build(target: str = "weixin") -> str:
    """执行项目构建。target 可选 weixin、alipay、h5、tt、baidu。"""
    root = workspace_root()
    pkg = _read_package_json(root)
    if pkg is None:
        return tool_error("未找到 package.json，无法推断构建命令")

    scripts = pkg.get("scripts") or {}
    if not isinstance(scripts, dict):
        return tool_error("package.json scripts 无效")

    target = target.strip().lower() or "weixin"
    script_name = _pick_build_script(scripts, target)
    if script_name is None:
        return tool_error(
            f"未找到适合 target={target!r} 的构建脚本",
            available=list(scripts.keys()),
            supported_targets=list(TARGET_SCRIPTS),
        )

    try:
        proc = _run_script(root, script_name)
        output = (proc.stdout or "") + (proc.stderr or "")
        tail = output[-BUILD_OUTPUT_TAIL:] if len(output) > BUILD_OUTPUT_TAIL else output
        if proc.returncode != 0:
            return tool_error(
                f"构建失败: {_package_manager(root)} run {script_name}",
                exit_code=proc.returncode,
                output=tail,
            )
        return tool_ok(
            script=script_name,
            target=target,
            package_manager=_package_manager(root),
            output=tail,
        )
    except subprocess.TimeoutExpired:
        return tool_error(f"构建超时（>{BUILD_TIMEOUT_SEC}s）: {script_name}")
    except FileNotFoundError:
        return tool_error(f"未找到包管理器: {_package_manager(root)}")
    except Exception as exc:
        return tool_error(str(exc))

File: apps/server/src/tools/project.ts
import { execFile } from "node:child_process";
import fs from "node:fs";
import path from "node:path";
import { promisify } from "node:util";

import { registerTool } from "./registry.js";
import { toolError, toolOk, workspaceRoot } from "./workspace.js";

const execFileAsync = promisify(execFile);

const BUILD_TIMEOUT_SEC = 600;
const BUILD_OUTPUT_TAIL = 4000;

const TARGET_SCRIPTS: Record<string, string[]> = {
  weixin: ["build:weapp", "build:mp-weixin", "build:weixin", "dev:mp-weixin"],
  alipay: ["build:alipay", "build:mp-alipay", "dev:mp-alipay"],
  h5: ["build:h5", "build:web", "build"],
  tt: ["build:tt", "build:mp-toutiao"],
  baidu: ["build:swan", "build:mp-baidu"],
};

function readPackageJson(root: string): Record<string, unknown> | null {
  const pkg = path.join(root, "package.json");
  if (!fs.existsSync(pkg)) return null;
  try {
    const data = JSON.parse(fs.readFileSync(pkg, "utf8"));
    return typeof data === "object" && data ? (data as Record<string, unknown>) : null;
  } catch {
    return null;
  }
}

function collectDeps(pkg: Record<string, unknown>): Record<string, string> {
  const deps: Record<string, string> = {};
  for (const key of ["dependencies", "devDependencies", "peerDependencies"]) {
    const raw = pkg[key];
    if (raw && typeof raw === "object") {
      Object.assign(deps, raw as Record<string, string>);
    }
  }
  return deps;
}

function detectSignals(root: string, deps: Record<string, string>): string[] {
  const signals: string[] = [];
  if (Object.keys(deps).some((k) => k.startsWith("@tarojs/"))) signals.push("taro");
  if (Object.keys(deps).some((k) => k.startsWith("@dcloudio/") || k.includes("uni-app"))) {
    signals.push("uni-app");
  }
  if (fs.existsSync(path.join(root, "pages.json")) || fs.existsSync(path.join(root, "manifest.json"))) {
    if (!signals.includes("uni-app")) signals.push("uni-app-like");
  }
  if (Object.keys(deps).some((k) => k.toLowerCase().includes("morjs")) || fs.existsSync(path.join(root, "mor.config.js"))) {
    signals.push("morjs");
  }
  if (fs.existsSync(path.join(root, "project.config.json"))) {
    signals.push("wechat-miniprogram-config");
  }
  return signals;
}

function packageManager(root: string): string {
  if (fs.existsSync(path.join(root, "pnpm-lock.yaml"))) return "pnpm";
  if (fs.existsSync(path.join(root, "yarn.lock"))) return "yarn";
  return "npm";
}

function pickBuildScript(scripts: Record<string, string>, target: string): string | null {
  for (const name of [...(TARGET_SCRIPTS[target] ?? []), "build"]) {
    if (name in scripts) return name;
  }
  return null;
}

async function runScript(root: string, scriptName: string) {
  const pm = packageManager(root);
  const cmd =
    pm === "pnpm"
      ? ["pnpm", "run", scriptName]
      : pm === "yarn"
        ? ["yarn", scriptName]
        : ["npm", "run", scriptName];
  return execFileAsync(cmd[0], cmd.slice(1), {
    cwd: root,
    timeout: BUILD_TIMEOUT_SEC * 1000,
    maxBuffer: 20 * 1024 * 1024,
  });
}

registerTool({
  name: "detect_framework",
  risk: "read",
  description: "检测工作区项目使用的多端框架（uni-app/Taro/Morjs 等）及可用构建脚本。",
  parameters: { type: "object", properties: {} },
  handler: async () => {
    const root = workspaceRoot();
    const pkg = readPackageJson(root);
    if (!pkg) return toolError("未找到可解析的 package.json");
    const deps = collectDeps(pkg);
    const signals = detectSignals(root, deps);
    const scripts = (pkg.scripts as Record<string, string>) ?? {};
    let framework = "unknown";
    if (signals.includes("taro")) framework = "taro";
    else if (signals.includes("uni-app") || signals.includes("uni-app-like")) framework = "uni-app";
    else if (signals.includes("morjs")) framework = "morjs";
    else if (signals.length) framework = signals[0];
    return toolOk({
      framework,
      signals,
      name: pkg.name,
      package_manager: packageManager(root),
      scripts: Object.keys(scripts),
      suggested_build: Object.fromEntries(
        Object.keys(TARGET_SCRIPTS).map((t) => [t, pickBuildScript(scripts, t)]),
      ),
    });
  },
});

registerTool({
  name: "run_build",
  risk: "exec",
  description: "执行项目构建。target 可选 weixin、alipay、h5、tt、baidu。",
  parameters: {
    type: "object",
    properties: { target: { type: "string" } },
  },
  handler: async (args) => {
    const root = workspaceRoot();
    const pkg = readPackageJson(root);
    if (!pkg) return toolError("未找到 package.json，无法推断构建命令");
    const scripts = (pkg.scripts as Record<string, string>) ?? {};
    const target = String(args.target ?? "weixin").trim().toLowerCase() || "weixin";
    const scriptName = pickBuildScript(scripts, target);
    if (!scriptName) {
      return toolError(`未找到适合 target=${target} 的构建脚本`, {
        available: Object.keys(scripts),
        supported_targets: Object.keys(TARGET_SCRIPTS),
      });
    }
    try {
      const { stdout, stderr } = await runScript(root, scriptName);
      const output = (stdout ?? "") + (stderr ?? "");
      const tail =
        output.length > BUILD_OUTPUT_TAIL ? output.slice(-BUILD_OUTPUT_TAIL) : output;
      return toolOk({
        script: scriptName,
        target,
        package_manager: packageManager(root),
        output: tail,
      });
    } catch (err: unknown) {
      const e = err as { stdout?: string; stderr?: string; code?: number; message?: string };
      const output = (e.stdout ?? "") + (e.stderr ?? "");
      const tail =
        output.length > BUILD_OUTPUT_TAIL ? output.slice(-BUILD_OUTPUT_TAIL) : output;
      return toolError(`构建失败: ${packageManager(root)} run ${scriptName}`, {
        exit_code: e.code,
        output: tail,
      });
    }
  },
});
